src/torch_ucc.cpp

```diff
@@ -305,7 +305,9 @@ void CommPG::ucx_disconnect_eps(
   if ((size_t)oob->store->add(oob->getKey("epclosed"), 1) == eps.size()) {
     oob->store->add(oob->getKey("epfinished"), 1);
   } else {
-    oob->store->wait({oob->getKey("epfinished")});
+    while (!oob->store->check({oob->getKey("epfinished")})) {
+      ucp_worker_progress(ucx_comm.worker);
```

Comment:
a check() merely will check if the key is present or not - for a counter-based key - this will always result in finding the key - esp if the check() is after the add() - 
If the intent here is to keep driving ucp_worker_progerss() until all the ranks have incremented the counter, what we should really be doing here is - 
  while ((size_t)oob->store->add(oob->getKey("epfinished"), 0) != eps.size()) {
     ucp_worker_progress(ucx_comm.worker);
     // add backoff
     std::this_thread::sleep_for(std::chrono::milliseconds(kSleepDuration));
  }

but the actual hang is not here - the actual hang happens in the previous do..while loop between lines:298..301

Reply:
@pallab-zz - thanks for the review! This makes sense, I will try `add()` and the sleep/spin. 

> but the actual hang is not here - the actual hang happens in the previous do..while loop between lines:298..301

I'm not sure if we're experiencing the same hang. In my case, I see one rank stuck in lines:298..301 like you said and other ranks stuck in `wait()` (i.e., before this change). When I do `check()` + `ucp_worker_progress()` instead of just `wait()`, I see other ranks enter `ucp_worker_progress()` (i.e., `check()` returns false even after `add()`) and allow the rank in lines:298..301 to progress and finish properly.

Reply:
check() will not block if the key is added - which is already added - so may be waiting for all of these is not even needed. and the reason the other rank may appear to finish due to this is likely because of end-of-pipe (since the checking rank excited) - otherwise i do not see how a different process can unstuck another process